tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwsHelper.java

```diff
@@ -0,0 +1,191 @@
+package org.jboss.eap.qe.microprofile.jwt.auth.tool;
+
+import java.net.URI;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.util.List;
+import java.util.Map;
+
+import org.jose4j.jwa.AlgorithmConstraints;
+import org.jose4j.jws.AlgorithmIdentifiers;
+import org.jose4j.jws.JsonWebSignature;
+import org.jose4j.jwt.consumer.JwtConsumer;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+import org.jose4j.jwt.consumer.JwtContext;
+
+/**
+ * Utility wrapper for the jose4j library for creating and reading JWS (JWT Signed Tokens)
+ * @author tborgato
+ */
+public class JwsHelper {
+
+    private RsaKeyHelper rsaHelper;
+    private EcKeyHelper ecHelper;
+
+    private RsaKeyHelper getRsaHelper() {
+        if (rsaHelper == null)
+            rsaHelper = new RsaKeyHelper();
+        return rsaHelper;
+    }
+
+    private EcKeyHelper getEcHelper() {
+        if (ecHelper == null)
+            ecHelper = new EcKeyHelper();
+        return ecHelper;
+    }
+
+    /**
+     * Generates a signed JWT
+     *
+     * @param algorithm
+     * @param privateKey
+     * @param kid
+     * @param issuer
+     * @param audience
+     * @param subject
+     * @param groups
+     * @param customClaims
+     * @return
+     * @throws Exception
+     */
+    public JsonWebSignature generateProperSignedJwt(
```

Comment:
What is the reason for this method when `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims` exist?

Reply:
I can't say exactly because it was there before my modifications .... I'm responsible for JwsHelper#generateProperSignedJwt which basically emulates JwtHelper#generateProperSignedJwtWithClaims but using `jose4j`

Reply:
@tommaso-borgato I understand it uses `jose4j` but why it is used here now?

Reply:
@honza-kasik it's the counterpart of the former method `generateProperSignedJwtWithClaims` but using `jose4j`

Reply:
@tommaso-borgato Sorry, I might worded that wrong. What I meant is that it is obvious jose4j is used. But if we are going to introduce it, it should be documented why.

Reply:
@honza-kasik I deprecated `generateProperSignedJwtWithClaims` and added some documentation to justify it

Reply:
@tommaso-borgato IMHO we are still missing docs why jose4j was added (the reason for deprecation)

Reply:
@honza-kasik just point me to the correct location for adding that .... and I'll add the docs

Reply:
basically it' here https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/release-notes.asciidoc#release-notes-for-microprofile-jwt-12:
```
TCK tests now use Jose4J to sign and encrypt the tokens.
```

Reply:
@tommaso-borgato From point of view of someone who just opened testsuite it looks like this:

* There's this and `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims`
* `org.jboss.eap.qe.microprofile.jwt.auth.tool.JwtHelper#generateProperSignedJwtWithClaims` is marked as deprecated and jose4j is mentioned
* Explanation how is jose4j reason for deprecation os nowhere to be found.

> TCK tests now use Jose4J to sign and encrypt the tokens.

We migrated some TCK tests here?

Reply:
`Jose4J` makes working with JWT stuff easier and alleviates the burden of rewriting and maintaining those functionalities .... just like any other library

Reply:
@honza-kasik is the javadoc https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/08968f5d3bab6a3f30d2a77606921a883309bd41/tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwtHelper.java#L137 enough or shall any more refactoring is needed?

Reply:
@honza-kasik I added the info you added here https://github.com/jboss-eap-qe/eap-microprofile-test-suite/blob/32a6fd6b335be4aeec2d8d0d38c7aba8c429297a/tooling-mp-jwt-auth-tool/src/main/java/org/jboss/eap/qe/microprofile/jwt/auth/tool/JwtHelper.java#L138; I added the things actually used used, let me know if it's enough or if I need to add some more extensive list

Reply:
Thank you @tommaso-borgato regarding documentation this is sufficient now.

However the introduced changes still bring a lot of redundancy. Since the re-write would mean a big investment and possible planning issues it will be merged with the redundant code present. Optimal solution would mean extending existing tooling with missing features such as JWT encryption and elliptic curve support. If this will present increased maintenance cost in the future. It can be "fixed" later.